platform/MCU/shared/STM32/src/hw_ticks.c

```diff
@@ -0,0 +1,85 @@
+/**
+  Copyright (c) 2013-2015 Particle Industries, Inc.  All rights reserved.
+
+  Copyright 2012 STMicroelectronics
+  http://www.st.com/software_license_agreement_liberty_v2
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation, either
+  version 3 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ ******************************************************************************
+ */
+
+
+#include "hw_ticks.h"
+#include <limits.h>
+
+/**
+ * The current millisecond counter value.
+ */
+static volatile system_tick_t system_millis = 0;
+
+/**
+ * The system clock value at last time system_millis was updated. This is updated
+ * after system_millis is incremented.
+ */
+static volatile system_tick_t system_millis_clock = 0;
+
+/**
+ * Increment the millisecond tick counter.
+ */
+void System1MsTick(void)
+{
+    system_millis++;
+    system_millis_clock = DWT->CYCCNT;
```

Comment:
You may want to comment here that the order of these two statements are critical. If someone changes them in the future, the clever trick below that ensures that you weren't interrupted between these two statements would fail.